Setup unit testing bypassing REST

lampajr opened the pull request:
<!-- If your PR fixes an open issue, use `Closes #435` to link your PR with the issue. #435 stands for the issue number you are fixing -->

## Changes proposed

With this PR I am proposing teh following improvements:
- [x] Setup new tests bypassing the REST, this way we can directly inject and call the services we want to test (I added them for `SchemaService` and `TestService`).
- [x] Improve test coverage for `TestService`, this won't be a huge improvement but at least a starting point
   * Before:
![TestServiceCoverageBefore](https://github.com/Hyperfoil/Horreum/assets/26715795/1a186146-f2b0-47a4-93f8-8fcc9b741792)
   * After:  
![TestServiceCoverageAfter](https://github.com/Hyperfoil/Horreum/assets/26715795/817b37b6-7396-49c5-8789-dab719795296)

- [x] Make some fixes or improvements that I noticed while implementing additional tests.

## Check List (Check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

lampajr commented:
Unfortunately, I noticed only right now that I introduced some formatting changes :disappointed:

johnaohara commented on horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/SchemaServiceImpl.java:
```diff
@@ -683,16 +688,17 @@ public Integer addOrUpdateLabel(int schemaId, Label labelDTO) {
-            label.id = -1;
-            existing = label;
-         }
+         LabelDAO existing = (LabelDAO) LabelDAO.findByIdOptional(label.id)
+               .orElseThrow(() -> ServiceException.notFound("Label " + label.id + " not found"));
+
```
This changes behaviour slightly, before if "existing" was not found, it was set to label, and label.idset to -1

lampajr replied:
That's true, but actually that case was not working at all. 

It was setting `existing = label` and then Hibernate would have thrown an exception as the entity it was trying to persist was not managed/associated to the Hibernate session (I think because the id should not be set to `-1` but to `null`).

Given that it was not working before, IMO to be coherent wrt other services if the provided id is not null and there is no entity with that id we should simply fail. Otherwise if we have user wanting to update a label, set the wrong id, it would end up creating a new label which is not the intended behavior. wdyt?

johnaohara commented on horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/SchemaServiceImpl.java:
```diff
@@ -272,13 +270,19 @@ public void dropToken(int id) {
    }
 
    public String updateToken(int id, String token) {
-      Query query = em.createNativeQuery(UPDATE_TOKEN);
```
I am +100 for removing native queries, we lose type safety in the code and couple ourselves to the db schema with every native query we have, esp for simple cases like this

lampajr replied:
Good to hear that, I think there are many other places (in other services) where we could remove them. I will try do that as part of other testing improvements :rocket:

johnaohara commented on horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:
```diff
@@ -157,9 +154,10 @@ public Test getByNameOrId(String input){
       TestDAO test = null;
       if (input.matches("-?\\d+")) {
          int id = Integer.parseInt(input);
+         // there could be some issue if name is numeric and corresponds to another test id
```
Can you open a new issue for this?

lampajr replied:
Created --> https://github.com/Hyperfoil/Horreum/issues/1670

johnaohara commented on horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:
```diff
@@ -471,37 +476,36 @@ public void dropToken(int testId, int tokenId) {
    }
 
    @Override
    @RolesAllowed("tester")
```
Missed another `@RolesAllowed(Roles.TESTER)` here

lampajr replied:
Good catch, thx!

johnaohara commented on horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/SchemaServiceNoRestTest.java:
```diff
@@ -0,0 +1,639 @@
+@TestProfile(HorreumTestProfile.class)
+@TestTransaction
+@TestSecurity(user = DEFAULT_USER, roles = {Roles.TESTER, Roles.VIEWER, FOO_TEAM, FOO_TESTER})
+class SchemaServiceNoRestTest extends BaseServiceNoRestTest {
```
For the tests that expect `ServiceExceptions`, I think we should check the response code to validate that the service is sending the correct http reponse as well as the error message

lampajr replied:
I agree, going to add those checks :+1:

johnaohara replied:
@lampajr have you added these checks?

lampajr replied:
yes, see as an example: https://github.com/Hyperfoil/Horreum/pull/1656/files#diff-8f4fde724167cf6f29ef5a003807126facd504f1158a9cd6462c1be599242d81R116

johnaohara commented on horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/SchemaServiceNoRestTest.java:
```diff
@@ -0,0 +1,639 @@
+      Schema schema = createSampleSchema("Dummy schema", schemaUri, FOO_TEAM);
+
+      ServiceException thrown = assertThrows(ServiceException.class, () -> schemaService.add(schema));
+      assertEquals("Please use URI starting with one of these schemes: [urn, uri, http, https, ftp, file, jar]", thrown.getMessage());
```
In relation to the previous comment about the Exception message, something like 

```suggestion
      assertEquals(Response.Status.BAD_REQUEST.getStatusCode();, thrown.getResponse().getStatus());
      assertEquals("Please use URI starting with one of these schemes: [urn, uri, http, https, ftp, file, jar]", thrown.getMessage());
```

lampajr replied:
Added for all tests that were checking the `ServiceException`

johnaohara replied:
great, I can see the changes now, might have been looking at an old version of the files :/

johnaohara commented on horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/SchemaServiceNoRestTest.java:
```diff
@@ -0,0 +1,639 @@
+   }
+
+   // utility to create a schema in the db, tested with testCreateSchema
+   private Schema createSchema(String name, String uri) {
+      return createSchema(name, uri, null);
+   }
+
+   // utility to create a schema in the db, tested with testCreateSchema
+   private Schema createSchema(String name, String uri, String token) {
+      Schema schema = new Schema();
+      schema.owner = FOO_TEAM;
+      schema.name = name;
+      schema.uri = uri;
+      schema.token = token;
+      schema.id = schemaService.add(schema);
+      return schema;
+   }
```
What is the difference between this and `BaseServiceNoRestTest.createSampleSchema()`?
Instead of duplicating code, we should use the method in the parent class

lampajr replied:
The only difference is line `636`, `createSampleSchema` simply creates the DTO object.. whereas `createSchema` creates a DTO and call the service to add it to the db.

I am refactoring this to make use of `createSampleSchema` inside `createSchema` body, something like:
```java
   // utility to create a schema in the db, tested with testCreateSchema
   private Schema createSchema(String name, String uri, String token) {
      Schema schema = createSampleSchema(name, uri, FOO_TEAM, token);
      schema.id = schemaService.add(schema);
      return schema;
   }
```

johnaohara commented on horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/TestServiceNoRestTest.java:
```diff
@@ -0,0 +1,504 @@
+      Test t = createSampleTest("", null, null, null);
+
+      // empty test name
+      ServiceException thrown = assertThrows(ServiceException.class, () -> testService.add(t));
+      assertEquals("Test name can not be empty", thrown.getMessage());
```
Same comment I made in `SchemaServiceNoRestTest` about testing the Response http code against expected code

lampajr commented:
Hi @johnaohara , thanks for the review :pray: 
I think I addressed all your comments (and applied changes accordingly) - see also this https://github.com/Hyperfoil/Horreum/pull/1656#discussion_r1595321834

lampajr commented:
> @lampajr please can you squash and then I will merge, thanks

Done :heavy_check_mark:
